diff --git a/content/browser/renderer_host/legacy_render_widget_host_win.cc b/content/browser/renderer_host/legacy_render_widget_host_win.cc
index e63c2f637828d..aa46e93035953 100644
--- a/content/browser/renderer_host/legacy_render_widget_host_win.cc
+++ b/content/browser/renderer_host/legacy_render_widget_host_win.cc
@@ -181,6 +181,10 @@ bool LegacyRenderWidgetHostHWND::InitOrDeleteSelf(HWND parent) {
   }
   ax_system_caret_ = std::move(ax_system_caret);
 
+  // Only register a touch window if we are using WM_TOUCH.
+  if (!features::IsUsingWMPointerForTouch())
+    RegisterTouchWindow(hwnd(), TWF_WANTPALM);
+
   // Ignore failure from this call. Some SKUs of Windows such as Hololens do not
   // support MSAA, and this call failing should not stop us from initializing
   // UI Automation support.
diff --git a/ui/base/ui_base_features.cc b/ui/base/ui_base_features.cc
index 6a4eaa26c90a9..ae5fd959e080b 100644
--- a/ui/base/ui_base_features.cc
+++ b/ui/base/ui_base_features.cc
@@ -281,6 +281,17 @@ BASE_FEATURE(kFocusFollowsCursor,
 // Enables focus follow follow cursor (sloppyfocus).
 BASE_FEATURE(kFocusFollowsCursor, base::FEATURE_DISABLED_BY_DEFAULT);
 
+#if BUILDFLAG(IS_WIN)
+// Enables using WM_POINTER instead of WM_TOUCH for touch events.
+BASE_FEATURE(kPointerEventsForTouch,
+             base::FEATURE_ENABLED_BY_DEFAULT);
+
+bool IsUsingWMPointerForTouch() {
+  return base::FeatureList::IsEnabled(kPointerEventsForTouch);
+}
+
+#endif  // BUILDFLAG(IS_WIN)
+
 BASE_FEATURE(kDragDropOnlySynthesizeHttpOrHttpsUrlsFromText,
              base::FEATURE_ENABLED_BY_DEFAULT);
 
diff --git a/ui/base/ui_base_features.h b/ui/base/ui_base_features.h
index 6182776562a8b..eb6ff594c79b3 100644
--- a/ui/base/ui_base_features.h
+++ b/ui/base/ui_base_features.h
@@ -47,7 +47,11 @@ COMPONENT_EXPORT(UI_BASE_FEATURES)
 #if BUILDFLAG(IS_WIN)
 COMPONENT_EXPORT(UI_BASE_FEATURES)
 BASE_DECLARE_FEATURE(kCalculateNativeWinOcclusion);
+COMPONENT_EXPORT(UI_BASE_FEATURES) BASE_DECLARE_FEATURE(kPointerEventsForTouch);
 
+// Returns true if the system should use WM_POINTER events for touch events.
+COMPONENT_EXPORT(UI_BASE_FEATURES) bool IsUsingWMPointerForTouch();
+
 COMPONENT_EXPORT(UI_BASE_FEATURES)
 BASE_DECLARE_FEATURE(kApplyNativeOcclusionToCompositor);
 COMPONENT_EXPORT(UI_BASE_FEATURES)
diff --git a/ui/views/win/hwnd_message_handler.cc b/ui/views/win/hwnd_message_handler.cc
index 509bbf34ed99b..328cf96499fa2 100644
--- a/ui/views/win/hwnd_message_handler.cc
+++ b/ui/views/win/hwnd_message_handler.cc
@@ -426,7 +426,8 @@ HWNDMessageHandler::HWNDMessageHandler(HWNDMessageHandlerDelegate* delegate,
       dwm_transition_desired_(false),
       sent_window_size_changing_(false),
       left_button_down_on_caption_(false),
-      background_fullscreen_hack_(false) {}
+      background_fullscreen_hack_(false),
+      pointer_events_for_touch_(::features::IsUsingWMPointerForTouch()) {}
 
 HWNDMessageHandler::~HWNDMessageHandler() {
   // Unhook MoveLoopMouseWatcher, to prevent call backs to this after deletion.
@@ -1742,6 +1743,9 @@ LRESULT HWNDMessageHandler::OnCreate(CREATESTRUCT* create_struct) {
   // Get access to a modifiable copy of the system menu.
   ::GetSystemMenu(hwnd(), false);
 
+  if (!pointer_events_for_touch_)
+    RegisterTouchWindow(hwnd(), TWF_WANTPALM);
+
   // We need to allow the delegate to size its contents since the window may not
   // receive a size notification when its initial bounds are specified at window
   // creation time.
@@ -2134,7 +2138,11 @@ LRESULT HWNDMessageHandler::OnPointerEvent(UINT message,
     case PT_PEN:
       return HandlePointerEventTypePenClient(message, w_param, l_param);
     case PT_TOUCH:
-      return HandlePointerEventTypeTouchOrNonClient(message, w_param, l_param);
+      if (pointer_events_for_touch_) {
+        return HandlePointerEventTypeTouchOrNonClient(
+            message, w_param, l_param);
+      }
+      [[fallthrough]];
     default:
       break;
   }
@@ -2785,11 +2793,96 @@ void HWNDMessageHandler::OnTimeChange() {
 LRESULT HWNDMessageHandler::OnTouchEvent(UINT message,
                                          WPARAM w_param,
                                          LPARAM l_param) {
-  // Release any associated memory with this event.
-  CloseTouchInputHandle(reinterpret_cast<HTOUCHINPUT>(l_param));
+  if (pointer_events_for_touch_) {
+    // Release any associated memory with this event.
+    CloseTouchInputHandle(reinterpret_cast<HTOUCHINPUT>(l_param));
+
+    // Claim the event is handled. This shouldn't ever happen
+    // because we don't register touch windows when we are using
+    // pointer events.
+    return 0;
+  }
+
+  // Handle touch events only on Aura for now.
+  WORD num_points = LOWORD(w_param);
+  base::HeapArray<TOUCHINPUT> input =
+      base::HeapArray<TOUCHINPUT>::WithSize(num_points);
+  if (ui::GetTouchInputInfoWrapper(reinterpret_cast<HTOUCHINPUT>(l_param),
+                                   num_points, input.data(),
+                                   sizeof(TOUCHINPUT))) {
+    // input[i].dwTime doesn't necessarily relate to the system time at all,
+    // so use base::TimeTicks::Now().
+    const base::TimeTicks event_time = base::TimeTicks::Now();
+    TouchEvents touch_events;
+    TouchIDs stale_touches(touch_ids_);
+
+    for (size_t i = 0; i < num_points; ++i) {
+      stale_touches.erase(input[i].dwID);
+      POINT point;
+      point.x = TOUCH_COORD_TO_PIXEL(input[i].x);
+      point.y = TOUCH_COORD_TO_PIXEL(input[i].y);
+      ScreenToClient(hwnd(), &point);
+
+      last_touch_or_pen_message_time_ = ::GetMessageTime();
+
+      gfx::Point touch_point(point.x, point.y);
+      auto touch_id = static_cast<ui::PointerId>(
+          id_generator_.GetGeneratedID(input[i].dwID));
+
+      if (input[i].dwFlags & TOUCHEVENTF_DOWN) {
+        touch_ids_.insert(input[i].dwID);
+        GenerateTouchEvent(ui::EventType::kTouchPressed, touch_point, touch_id,
+                           event_time, &touch_events);
+        ui::ComputeEventLatencyOSFromTOUCHINPUT(ui::EventType::kTouchPressed,
+                                                input[i], event_time);
+        touch_down_contexts_++;
+        base::SingleThreadTaskRunner::GetCurrentDefault()->PostDelayedTask(
+            FROM_HERE,
+            base::BindOnce(&HWNDMessageHandler::ResetTouchDownContext,
+                           msg_handler_weak_factory_.GetWeakPtr()),
+            kTouchDownContextResetTimeout);
+      } else {
+        if (input[i].dwFlags & TOUCHEVENTF_MOVE) {
+          GenerateTouchEvent(ui::EventType::kTouchMoved, touch_point, touch_id,
+                             event_time, &touch_events);
+          ui::ComputeEventLatencyOSFromTOUCHINPUT(ui::EventType::kTouchMoved,
+                                                  input[i], event_time);
+        }
 
-  // Claim the event is handled. This shouldn't ever happen because we don't
-  // register touch windows when we are using pointer events.
+        if (input[i].dwFlags & TOUCHEVENTF_UP) {
+          touch_ids_.erase(input[i].dwID);
+          GenerateTouchEvent(ui::EventType::kTouchReleased, touch_point,
+                             touch_id, event_time, &touch_events);
+          ui::ComputeEventLatencyOSFromTOUCHINPUT(ui::EventType::kTouchReleased,
+                                                  input[i], event_time);
+          id_generator_.ReleaseNumber(input[i].dwID);
+        }
+      }
+    }
+    // If a touch has been dropped from the list (without a TOUCH_EVENTF_UP)
+    // we generate a simulated TOUCHEVENTF_UP event.
+    for (auto touch_number : stale_touches) {
+      // Log that we've hit this code. When usage drops off, we can remove
+      // this "workaround". See https://crbug.com/811273
+      UMA_HISTOGRAM_BOOLEAN("TouchScreen.MissedTOUCHEVENTF_UP", true);
+      auto touch_id = static_cast<ui::PointerId>(
+          id_generator_.GetGeneratedID(touch_number));
+      touch_ids_.erase(touch_number);
+      GenerateTouchEvent(ui::EventType::kTouchReleased, gfx::Point(0, 0),
+                         touch_id, event_time, &touch_events);
+      id_generator_.ReleaseNumber(touch_number);
+    }
+
+    // Handle the touch events asynchronously. We need this because touch
+    // events on windows don't fire if we enter a modal loop in the context of
+    // a touch event.
+    base::SingleThreadTaskRunner::GetCurrentDefault()->PostTask(
+        FROM_HERE,
+        base::BindOnce(&HWNDMessageHandler::HandleTouchEvents,
+                       msg_handler_weak_factory_.GetWeakPtr(), touch_events));
+  }
+  CloseTouchInputHandle(reinterpret_cast<HTOUCHINPUT>(l_param));
+  SetMsgHandled(FALSE);
   return 0;
 }
 
@@ -3668,6 +3668,22 @@ void HWNDMessageHandler::UpdateDwmFrame() {
   }
 }
 
+void HWNDMessageHandler::GenerateTouchEvent(ui::EventType event_type,
+                                            const gfx::Point& point,
+                                            ui::PointerId id,
+                                            base::TimeTicks time_stamp,
+                                            TouchEvents* touch_events) {
+  ui::TouchEvent event(event_type, point, time_stamp,
+                       ui::PointerDetails(ui::EventPointerType::kTouch, id));
+
+  event.SetFlags(ui::GetModifiersFromKeyState());
+
+  event.latency()->AddLatencyNumberWithTimestamp(
+      ui::INPUT_EVENT_LATENCY_ORIGINAL_COMPONENT, time_stamp);
+
+  touch_events->push_back(event);
+}
+
 bool HWNDMessageHandler::HandleMouseInputForCaption(unsigned int message,
                                                     WPARAM w_param,
                                                     LPARAM l_param) {
diff --git a/ui/views/win/hwnd_message_handler.h b/ui/views/win/hwnd_message_handler.h
index 558f6ecaced6e..4299e81d4f9ed 100644
--- a/ui/views/win/hwnd_message_handler.h
+++ b/ui/views/win/hwnd_message_handler.h
@@ -609,6 +609,16 @@ class VIEWS_EXPORT HWNDMessageHandler : public gfx::WindowImpl,
   // Updates DWM frame to extend into client area if needed.
   void UpdateDwmFrame();
 
+  // Generates a touch event and adds it to the |touch_events| parameter.
+  // |point| is the point where the touch was initiated.
+  // |id| is the event id associated with the touch event.
+  // |time_stamp| is the time stamp associated with the message.
+  void GenerateTouchEvent(ui::EventType event_type,
+                          const gfx::Point& point,
+                          ui::PointerId id,
+                          base::TimeTicks time_stamp,
+                          TouchEvents* touch_events);
+
   // Handles WM_NCLBUTTONDOWN and WM_NCMOUSEMOVE messages on the caption.
   // Returns true if the message was handled.
   bool HandleMouseInputForCaption(unsigned int message,
@@ -690,6 +691,9 @@ class VIEWS_EXPORT HWNDMessageHandler : public gfx::WindowImpl,
   // area. We need this so we can correctly show the context menu on mouse-up.
   bool is_right_mouse_pressed_on_caption_;
 
+  // The set of touch devices currently down.
+  TouchIDs touch_ids_;
+
   // ScopedRedrawLock ----------------------------------------------------------
 
   // Represents the number of ScopedRedrawLocks active against this widget.
@@ -811,6 +814,10 @@ class VIEWS_EXPORT HWNDMessageHandler : public gfx::WindowImpl,
   // partially or fully transparent.
   bool is_translucent_ = false;
 
+  // True if the window should process WM_POINTER for touch events and
+  // not WM_TOUCH events.
+  bool pointer_events_for_touch_;
+
   // True if the window uses rounded corners. This uses DWMWCP_ROUND style on
   // Windows 11. Not effective on Windows 10.
   bool use_rounded_corner_ = false;
