diff --git a/content/child/font_data/font_data_manager.cc b/content/child/font_data/font_data_manager.cc
index 6b13c614e33ba..16212826d1d1a 100644
--- a/content/child/font_data/font_data_manager.cc
+++ b/content/child/font_data/font_data_manager.cc
@@ -27,12 +27,10 @@
 #include "third_party/blink/public/web/win/web_font_rendering.h"
 #include "third_party/skia/src/ports/SkTypeface_win_dw.h"  // nogncheck
 #endif
-#if !BUILDFLAG(IS_WIN)
 #if BUILDFLAG(ENABLE_FREETYPE)
 #include "third_party/skia/include/ports/SkFontMgr_empty.h"
 #endif
 #include "third_party/skia/include/ports/SkTypeface_fontations.h"
-#endif
 
 namespace font_data_service {
 
@@ -78,7 +76,7 @@ UNSAFE_BUFFER_USAGE std::vector<std::string> bcp47ArrayToVector(
 
 FontDataManager::FontDataManager()
     : typeface_cache_(kTypefaceCacheSize),
-#if !BUILDFLAG(IS_WIN) && BUILDFLAG(ENABLE_FREETYPE)
+#if BUILDFLAG(ENABLE_FREETYPE)
       custom_fnt_mgr_(SkFontMgr_New_Custom_Empty()),
 #endif
       main_task_runner_(base::SingleThreadTaskRunner::GetCurrentDefault()) {
@@ -267,11 +265,17 @@ sk_sp<SkTypeface> FontDataManager::onMakeFromStreamArgs(
     const SkFontArguments& args) const {
   TRACE_EVENT1("fonts", "FontDataManager::onMakeFromStreamArgs", "size",
                stream->getLength());
-#if BUILDFLAG(IS_WIN)
-  return DWriteFontTypeface::MakeFromStream(std::move(stream), args);
-#else
   // Experiment will test the performance of different SkTypefaces.
   // 'custom_fnt_mgr_' is a wrapper to create an SkFreeType typeface.
+
+  // DWRITE is only an option on Windows. Other platforms must use Freetype or
+  // Fontations.
+#if BUILDFLAG(IS_WIN)
+  if (features::kFontDataServiceTypefaceType.Get() ==
+      features::FontDataServiceTypefaceType::kDwrite) {
+    return DWriteFontTypeface::MakeFromStream(std::move(stream), args);
+  }
+#endif
   // Chromium currently always sets ENABLE_FREETYPE, but nonetheless allow
   // falling back to fontations if the param is set to freetype but freetype
   // isn't enabled.
@@ -283,7 +287,6 @@ sk_sp<SkTypeface> FontDataManager::onMakeFromStreamArgs(
 #endif
 
   return SkTypeface_Make_Fontations(std::move(stream), args);
-#endif
 }
 
 sk_sp<SkTypeface> FontDataManager::onMakeFromFile(const char path[],
diff --git a/content/child/font_data/font_data_manager.h b/content/child/font_data/font_data_manager.h
index f95ba5c3b118e..997f01c20cbd5 100644
--- a/content/child/font_data/font_data_manager.h
+++ b/content/child/font_data/font_data_manager.h
@@ -199,10 +199,7 @@ class CONTENT_EXPORT FontDataManager : public SkFontMgr,
   // onCountFamilies or onGetFamilyName.
   mutable std::vector<std::string> family_names_ GUARDED_BY(family_names_lock_);
 
-  // FreeType font manager used on non-Windows platforms (e.g. Linux/ChromeOS)
-  // when the FreeType backend is selected. Windows exclusively uses DirectWrite
-  // typefaces (via DWriteFontTypeface).
-#if !BUILDFLAG(IS_WIN) && BUILDFLAG(ENABLE_FREETYPE)
+#if BUILDFLAG(ENABLE_FREETYPE)
   sk_sp<SkFontMgr> custom_fnt_mgr_;
 #endif
   scoped_refptr<base::SingleThreadTaskRunner> main_task_runner_;
diff --git a/content/common/features.cc b/content/common/features.cc
index 7e7515685e7e7..434bd6d043af9 100644
--- a/content/common/features.cc
+++ b/content/common/features.cc
@@ -284,8 +284,22 @@ BASE_FEATURE(kFocusRenderWidgetHostViewAndroidOnMouseDown,
 #endif
 // This feature controls whether the renderer should use FontDataManager to
 // fetch fonts from the Browser's FontDataService. It is currently scoped to
-// Linux and ChromeOS. See
+// Windows and Linux (via separate features and experiments). See
 // crbug.com/335680565.
+#if BUILDFLAG(IS_WIN)
+BASE_FEATURE(kFontDataServiceAllWebContents, base::FEATURE_ENABLED_BY_DEFAULT);
+const base::FeatureParam<FontDataServiceTypefaceType>::Option
+    font_data_service_typeface[] = {
+        {FontDataServiceTypefaceType::kDwrite, "DWrite"},
+        {FontDataServiceTypefaceType::kFreetype, "Freetype"},
+        {FontDataServiceTypefaceType::kFontations, "Fontations"}};
+BASE_FEATURE_ENUM_PARAM(FontDataServiceTypefaceType,
+                        kFontDataServiceTypefaceType,
+                        &kFontDataServiceAllWebContents,
+                        "typeface",
+                        FontDataServiceTypefaceType::kDwrite,
+                        &font_data_service_typeface);
+#endif  // BUILDFLAG(IS_WIN)
 #if BUILDFLAG(IS_LINUX) || BUILDFLAG(IS_CHROMEOS)
 const base::FeatureParam<FontDataServiceTypefaceType>::Option
     font_data_service_typeface[] = {
@@ -320,7 +334,7 @@ BASE_FEATURE(kFontDataServiceForCSSLocalFonts,
 #if BUILDFLAG(IS_WIN) || BUILDFLAG(IS_LINUX) || BUILDFLAG(IS_CHROMEOS)
 bool IsFontDataServiceEnabled() {
 #if BUILDFLAG(IS_WIN)
-  return true;
+  return base::FeatureList::IsEnabled(features::kFontDataServiceAllWebContents);
 #elif BUILDFLAG(IS_LINUX)
   return base::FeatureList::IsEnabled(features::kFontDataServiceLinux);
 #elif BUILDFLAG(IS_CHROMEOS)
diff --git a/content/common/features.h b/content/common/features.h
index b5a77eecf2da6..151dd3d3771ad 100644
--- a/content/common/features.h
+++ b/content/common/features.h
@@ -105,6 +105,16 @@ CONTENT_EXPORT BASE_DECLARE_FEATURE(
     kFocusRenderWidgetHostViewAndroidOnMouseDown);
 #endif
 
+#if BUILDFLAG(IS_WIN)
+CONTENT_EXPORT BASE_DECLARE_FEATURE(kFontDataServiceAllWebContents);
+enum class FontDataServiceTypefaceType {
+  kDwrite,
+  kFreetype,
+  kFontations,
+};
+CONTENT_EXPORT BASE_DECLARE_FEATURE_PARAM(FontDataServiceTypefaceType,
+                                          kFontDataServiceTypefaceType);
+#endif  // BUILDFLAG(IS_WIN)
 #if BUILDFLAG(IS_LINUX) || BUILDFLAG(IS_CHROMEOS)
 #if BUILDFLAG(IS_LINUX)
 CONTENT_EXPORT BASE_DECLARE_FEATURE(kFontDataServiceLinux);
diff --git a/content/renderer/renderer_main_platform_delegate_win.cc b/content/renderer/renderer_main_platform_delegate_win.cc
index f5d81225df850..bfe2abf59aa59 100644
--- a/content/renderer/renderer_main_platform_delegate_win.cc
+++ b/content/renderer/renderer_main_platform_delegate_win.cc
@@ -15,6 +15,7 @@
 #include "base/win/windows_version.h"
 #include "content/child/dwrite_font_proxy/dwrite_font_proxy_init_impl_win.h"
 #include "content/child/font_warmup_win.h"
+#include "content/common/features.h"
 #include "content/public/common/content_features.h"
 #include "content/public/common/content_switches.h"
 #include "content/public/renderer/render_thread.h"
@@ -57,10 +58,13 @@ void RendererMainPlatformDelegate::PlatformInitialize() {
     std::unique_ptr<icu::TimeZone> zone(icu::TimeZone::createDefault());
   }
 
+  // Do not initialize DWriteFactory if the feature flag is enabled
+  // since this will conflict with the experimental font manager.
   // Fallback to only DWrite if running in single process mode, since there can
   // only be a single font manager and the browser process always has a DWrite
   // one.
-  if (command_line.HasSwitch(switches::kSingleProcess)) {
+  if (!base::FeatureList::IsEnabled(features::kFontDataServiceAllWebContents) ||
+      command_line.HasSwitch(switches::kSingleProcess)) {
     InitializeDWriteFontProxy();
   }
 }
