diff --git a/src/compiler/translator/hlsl/OutputHLSL.cpp b/src/compiler/translator/hlsl/OutputHLSL.cpp
index 0021413d95..39c19d63e3 100644
--- a/third_party/angle/src/compiler/translator/hlsl/OutputHLSL.cpp
+++ b/third_party/angle/src/compiler/translator/hlsl/OutputHLSL.cpp
@@ -957,21 +955,26 @@ void OutputHLSL::header(TInfoSinkBase &out,
                 out << "    float3 dx_DepthRange : packoffset(c0);\n";
             }
 
-            out << "    float2 dx_ViewScale  : packoffset(c1);\n";
+            // dx_ViewAdjust and dx_ViewCoords are only used by some shaders. However, we declare
+            // them for all shaders. The bytecode is the same whether we declare them or not, since
+            // D3DCompiler removes them if they're unused.
+            out << "    float4 dx_ViewAdjust : packoffset(c1);\n";
+            out << "    float2 dx_ViewCoords : packoffset(c2);\n";
+            out << "    float2 dx_ViewScale  : packoffset(c3);\n";
 
-            out << "    float clipControlOrigin : packoffset(c1.z);\n";
-            out << "    float clipControlZeroToOne : packoffset(c1.w);\n";
+            out << "    float clipControlOrigin : packoffset(c3.z);\n";
+            out << "    float clipControlZeroToOne : packoffset(c3.w);\n";
 
-            mResourcesHLSL->samplerMetadataUniforms(out, 3);
+            mResourcesHLSL->samplerMetadataUniforms(out, 5);
 
             if (mUsesVertexID)
             {
-                out << "    uint dx_VertexID : packoffset(c2.x);\n";
+                out << "    uint dx_VertexID : packoffset(c4.x);\n";
             }
 
             if (mClipDistanceSize)
             {
-                out << "    uint clipDistancesEnabled : packoffset(c2.y);\n";
+                out << "    uint clipDistancesEnabled : packoffset(c4.y);\n";
             }
 
             out << "};\n"
diff --git a/src/libANGLE/renderer/d3d/DynamicHLSL.cpp b/src/libANGLE/renderer/d3d/DynamicHLSL.cpp
index f7c64790e0..b3aac0882b 100644
--- a/third_party/angle/src/libANGLE/renderer/d3d/DynamicHLSL.cpp
+++ b/third_party/angle/src/libANGLE/renderer/d3d/DynamicHLSL.cpp
@@ -264,6 +264,7 @@ std::string DynamicHLSL::GenerateVertexShaderForInputLayout(
 
 // static
 std::string DynamicHLSL::GeneratePixelShaderForOutputSignature(
+    RendererD3D *renderer,
     const std::string &sourceShader,
     const std::vector<PixelShaderOutputVariable> &outputVariables,
     FragDepthUsage fragDepthUsage,
@@ -271,8 +272,13 @@ std::string DynamicHLSL::GeneratePixelShaderForOutputSignature(
     const std::vector<GLenum> &outputLayout,
     size_t baseUAVRegister)
 {
-    std::string targetSemantic = "SV_TARGET";
-    std::string depthSemantic  = [fragDepthUsage]() {
+    const int shaderModel      = renderer->getMajorShaderModel();
+    std::string targetSemantic = (shaderModel >= 4) ? "SV_TARGET" : "COLOR";
+    std::string depthSemantic  = [shaderModel, fragDepthUsage]() {
+        if (shaderModel < 4)
+        {
+            return "DEPTH";
+        }
         switch (fragDepthUsage)
         {
             case FragDepthUsage::Less:
@@ -292,6 +298,11 @@ std::string DynamicHLSL::GeneratePixelShaderForOutputSignature(
 
     size_t numOutputs = outputLayout.size();
 
+    // Workaround for HLSL 3.x: We can't do a depth/stencil only render, the runtime will complain.
+    if (numOutputs == 0 && shaderModel == 3)
+    {
+        numOutputs = 1u;
+    }
     const PixelShaderOutputVariable defaultOutput(GL_FLOAT_VEC4, "unused", "float4(0, 0, 0, 1)", 0,
                                                   0);
     size_t outputIndex = 0;
@@ -378,9 +389,11 @@ std::string DynamicHLSL::GenerateShaderForImage2DBindSignature(
 }
 
 // static
-void DynamicHLSL::GenerateVaryingLinkHLSL(const VaryingPacking &varyingPacking,
+void DynamicHLSL::GenerateVaryingLinkHLSL(RendererD3D *renderer,
+                                          const VaryingPacking &varyingPacking,
                                           const BuiltinInfo &builtins,
                                           FragDepthUsage fragDepthUsage,
+                                          bool programUsesPointSize,
                                           std::ostringstream &hlslStream)
 {
     ASSERT(builtins.dxPosition.enabled);
@@ -444,7 +457,8 @@ void DynamicHLSL::GenerateVaryingLinkHLSL(const VaryingPacking &varyingPacking,
                    << ";\n";
     }
 
-    std::string varyingSemantic = "TEXCOORD";
+    std::string varyingSemantic =
+        GetVaryingSemantic(renderer->getMajorShaderModel(), programUsesPointSize);
 
     const auto &registerInfos = varyingPacking.getRegisterList();
     for (GLuint registerIndex = 0u; registerIndex < registerInfos.size(); ++registerIndex)
@@ -508,6 +522,7 @@ void DynamicHLSL::GenerateVaryingLinkHLSL(const VaryingPacking &varyingPacking,
 
 // static
 void DynamicHLSL::GenerateShaderLinkHLSL(
+    RendererD3D *renderer,
     const gl::Caps &caps,
     const gl::ShaderMap<gl::SharedCompiledShaderState> &shaderData,
     const gl::ShaderMap<SharedCompiledShaderStateD3D> &shaderDataD3D,
@@ -522,9 +537,13 @@ void DynamicHLSL::GenerateShaderLinkHLSL(
 
     const gl::SharedCompiledShaderState &vertexShader   = shaderData[ShaderType::Vertex];
     const gl::SharedCompiledShaderState &fragmentShader = shaderData[ShaderType::Fragment];
+    const int shaderModel                               = renderer->getMajorShaderModel();
 
     const SharedCompiledShaderStateD3D &fragmentShaderD3D = shaderDataD3D[ShaderType::Fragment];
 
+    // usesViewScale() isn't supported in the D3D9 renderer
+    ASSERT(shaderModel >= 4 || !programMetadata.usesViewScale());
+
     // Validation done in the compiler
     ASSERT(!fragmentShaderD3D || !fragmentShaderD3D->usesFragColor ||
            !fragmentShaderD3D->usesFragData);
@@ -532,7 +551,8 @@ void DynamicHLSL::GenerateShaderLinkHLSL(
     std::ostringstream vertexStream;
     vertexStream << "struct VS_OUTPUT\n";
     const auto &vertexBuiltins = builtinsD3D[gl::ShaderType::Vertex];
-    GenerateVaryingLinkHLSL(varyingPacking, vertexBuiltins, FragDepthUsage::Unused, vertexStream);
+    GenerateVaryingLinkHLSL(renderer, varyingPacking, vertexBuiltins, FragDepthUsage::Unused,
+                            builtinsD3D.usesPointSize(), vertexStream);
 
     std::ostringstream vertexGenerateOutput;
     vertexGenerateOutput << "VS_OUTPUT generateOutput(VS_INPUT input)\n"
@@ -640,28 +660,52 @@ void DynamicHLSL::GenerateShaderLinkHLSL(
         vertexGenerateOutput << "    output.gl_Layer = ViewID_OVR;\n";
     }
 
-    vertexGenerateOutput << "    output.dx_Position.x = gl_Position.x;\n";
-
-    if (programMetadata.usesViewScale())
+    // On D3D9, we need to emulate large viewports using dx_ViewAdjust.
+    if (shaderModel >= 4)
     {
-        // This code assumes that dx_ViewScale.y = -1.0f when rendering to texture, and +1.0f
-        // when rendering to the default framebuffer. No other values are valid.
-        vertexGenerateOutput << "    output.dx_Position.y = dx_ViewScale.y * gl_Position.y;\n";
+        vertexGenerateOutput << "    output.dx_Position.x = gl_Position.x;\n";
+
+        if (programMetadata.usesViewScale())
+        {
+            // This code assumes that dx_ViewScale.y = -1.0f when rendering to texture, and +1.0f
+            // when rendering to the default framebuffer. No other values are valid.
+            vertexGenerateOutput << "    output.dx_Position.y = dx_ViewScale.y * gl_Position.y;\n";
+        }
+        else
+        {
+            vertexGenerateOutput
+                << "    output.dx_Position.y = clipControlOrigin * gl_Position.y;\n";
+        }
+
+        vertexGenerateOutput
+            << "    if (clipControlZeroToOne)\n"
+            << "    {\n"
+            << "        output.dx_Position.z = gl_Position.z;\n"
+            << "    } else {\n"
+            << "        output.dx_Position.z = (gl_Position.z + gl_Position.w) * 0.5;\n"
+            << "    }\n";
+
+        vertexGenerateOutput << "    output.dx_Position.w = gl_Position.w;\n";
     }
     else
     {
-        vertexGenerateOutput << "    output.dx_Position.y = clipControlOrigin * gl_Position.y;\n";
-    }
+        vertexGenerateOutput << "    output.dx_Position.x = gl_Position.x * dx_ViewAdjust.z + "
+                                "dx_ViewAdjust.x * gl_Position.w;\n";
 
-    vertexGenerateOutput
-        << "    if (clipControlZeroToOne)\n"
-        << "    {\n"
-        << "        output.dx_Position.z = gl_Position.z;\n"
-        << "    } else {\n"
-        << "        output.dx_Position.z = (gl_Position.z + gl_Position.w) * 0.5;\n"
-        << "    }\n";
+        vertexGenerateOutput << "    output.dx_Position.y = clipControlOrigin * (gl_Position.y "
+                                "* dx_ViewAdjust.w + "
+                                "dx_ViewAdjust.y * gl_Position.w);\n";
 
-    vertexGenerateOutput << "    output.dx_Position.w = gl_Position.w;\n";
+        vertexGenerateOutput
+            << "    if (clipControlZeroToOne)\n"
+            << "    {\n"
+            << "        output.dx_Position.z = gl_Position.z;\n"
+            << "    } else {\n"
+            << "        output.dx_Position.z = (gl_Position.z + gl_Position.w) * 0.5;\n"
+            << "    }\n";
+
+        vertexGenerateOutput << "    output.dx_Position.w = gl_Position.w;\n";
+    }
 
     // We don't need to output gl_PointSize if we use are emulating point sprites via instancing.
     if (vertexBuiltins.glPointSize.enabled)
@@ -732,7 +776,8 @@ void DynamicHLSL::GenerateShaderLinkHLSL(
 
     std::ostringstream pixelStream;
     pixelStream << "struct PS_INPUT\n";
-    GenerateVaryingLinkHLSL(varyingPacking, pixelBuiltins, programMetadata.getFragDepthUsage(),
+    GenerateVaryingLinkHLSL(renderer, varyingPacking, pixelBuiltins,
+                            programMetadata.getFragDepthUsage(), builtinsD3D.usesPointSize(),
                             pixelStream);
     pixelStream << "\n";
 
@@ -747,55 +792,89 @@ void DynamicHLSL::GenerateShaderLinkHLSL(
     {
         pixelPrologue << "    float rhw = 1.0 / input.gl_FragCoord.w;\n";
 
-        // gl_FragCoord is computed from dx_Position, which the pixel shader can read directly.
+        // Certain Shader Models (4_0+ and 3_0) allow reading from dx_Position in the pixel shader.
+        // Other Shader Models (2_x) don't support this, so we emulate it using
+        // dx_ViewCoords.
         // DComp usually gives us an offset at (0, 0), but this is not always the case. It is
         // valid for DComp to give us an offset into the texture atlas. In that scenario, we
         // need to offset gl_FragCoord using dx_FragCoordOffset to point to the correct location
         // of the pixel.
-        pixelPrologue << "    gl_FragCoord.x = input.dx_Position.x - dx_FragCoordOffset.x;\n"
-                      << "    gl_FragCoord.y = input.dx_Position.y - dx_FragCoordOffset.y;\n";
+        if (shaderModel >= 4)
+        {
+            pixelPrologue << "    gl_FragCoord.x = input.dx_Position.x - dx_FragCoordOffset.x;\n"
+                          << "    gl_FragCoord.y = input.dx_Position.y - dx_FragCoordOffset.y;\n";
+        }
+        else if (shaderModel == 3)
+        {
+            pixelPrologue
+                << "    gl_FragCoord.x = input.dx_Position.x + 0.5 - dx_FragCoordOffset.x;\n"
+                << "    gl_FragCoord.y = input.dx_Position.y + 0.5 - dx_FragCoordOffset.y;\n";
+        }
+        else
+        {
+            // dx_ViewCoords contains the viewport width/2, height/2, center.x and center.y. See
+            // Renderer::setViewport()
+            pixelPrologue
+                << "    gl_FragCoord.x = (input.gl_FragCoord.x * rhw) * dx_ViewCoords.x + "
+                   "dx_ViewCoords.z - dx_FragCoordOffset.x;\n"
+                << "    gl_FragCoord.y = (input.gl_FragCoord.y * rhw) * dx_ViewCoords.y + "
+                   "dx_ViewCoords.w - dx_FragCoordOffset.y;\n";
+        }
 
         if (programMetadata.usesViewScale())
         {
-            // Some assumptions:
-            //  - dx_ViewScale.y = -1.0f when rendering to texture
-            //  - dx_ViewScale.y = +1.0f when rendering to the default framebuffer
-            //  - gl_FragCoord.y has been set correctly above.
-            //
-            // When rendering to the backbuffer, the code inverts gl_FragCoord's y coordinate.
-            // This involves subtracting the y coordinate from the height of the area being
-            // rendered to.
-            //
-            // First we calculate the height of the area being rendered to:
-            //    render_area_height = (2.0f / (1.0f - input.gl_FragCoord.y * rhw)) *
-            //    gl_FragCoord.y
-            //
-            // Note that when we're rendering to default FB, we want our output to be
-            // equivalent to:
-            //    "gl_FragCoord.y = render_area_height - gl_FragCoord.y"
-            //
-            // When we're rendering to a texture, we want our output to be equivalent to:
-            //    "gl_FragCoord.y = gl_FragCoord.y;"
-            //
-            // If we set scale_factor = ((1.0f + dx_ViewScale.y) / 2.0f), then notice that
-            //  - When rendering to default FB: scale_factor = 1.0f
-            //  - When rendering to texture:    scale_factor = 0.0f
-            //
-            // Therefore, we can get our desired output by setting:
-            //    "gl_FragCoord.y = scale_factor * render_area_height - dx_ViewScale.y *
-            //    gl_FragCoord.y"
-            //
-            // Simplifying, this becomes:
-            pixelPrologue
-                << "    gl_FragCoord.y = (1.0f + dx_ViewScale.y) * gl_FragCoord.y /"
-                   "(1.0f - input.gl_FragCoord.y * rhw)  - dx_ViewScale.y * gl_FragCoord.y;\n";
+            // NOTE: usesViewScale() can only be true on D3D11 (i.e. Shader Model 4.0+).
+            if (shaderModel >= 4)
+            {
+                // Some assumptions:
+                //  - dx_ViewScale.y = -1.0f when rendering to texture
+                //  - dx_ViewScale.y = +1.0f when rendering to the default framebuffer
+                //  - gl_FragCoord.y has been set correctly above.
+                //
+                // When rendering to the backbuffer, the code inverts gl_FragCoord's y coordinate.
+                // This involves subtracting the y coordinate from the height of the area being
+                // rendered to.
+                //
+                // First we calculate the height of the area being rendered to:
+                //    render_area_height = (2.0f / (1.0f - input.gl_FragCoord.y * rhw)) *
+                //    gl_FragCoord.y
+                //
+                // Note that when we're rendering to default FB, we want our output to be
+                // equivalent to:
+                //    "gl_FragCoord.y = render_area_height - gl_FragCoord.y"
+                //
+                // When we're rendering to a texture, we want our output to be equivalent to:
+                //    "gl_FragCoord.y = gl_FragCoord.y;"
+                //
+                // If we set scale_factor = ((1.0f + dx_ViewScale.y) / 2.0f), then notice that
+                //  - When rendering to default FB: scale_factor = 1.0f
+                //  - When rendering to texture:    scale_factor = 0.0f
+                //
+                // Therefore, we can get our desired output by setting:
+                //    "gl_FragCoord.y = scale_factor * render_area_height - dx_ViewScale.y *
+                //    gl_FragCoord.y"
+                //
+                // Simplifying, this becomes:
+                pixelPrologue
+                    << "    gl_FragCoord.y = (1.0f + dx_ViewScale.y) * gl_FragCoord.y /"
+                       "(1.0f - input.gl_FragCoord.y * rhw)  - dx_ViewScale.y * gl_FragCoord.y;\n";
+            }
         }
 
-        pixelPrologue << "    gl_FragCoord.z = input.dx_Position.z;\n";
+        if (shaderModel >= 4)
+        {
+            pixelPrologue << "    gl_FragCoord.z = input.dx_Position.z;\n";
+        }
+        else
+        {
+            pixelPrologue
+                << "    gl_FragCoord.z = (input.gl_FragCoord.z * rhw) * dx_DepthFront.x + "
+                   "dx_DepthFront.y;\n";
+        }
         pixelPrologue << "    gl_FragCoord.w = rhw;\n";
     }
 
-    if (pixelBuiltins.glPointCoord.enabled)
+    if (pixelBuiltins.glPointCoord.enabled && shaderModel >= 3)
     {
         pixelPrologue << "    gl_PointCoord.x = input.gl_PointCoord.x;\n"
                       << "    gl_PointCoord.y = 1.0 - input.gl_PointCoord.y;\n";
@@ -803,7 +882,14 @@ void DynamicHLSL::GenerateShaderLinkHLSL(
 
     if (fragmentShaderD3D && fragmentShaderD3D->usesFrontFacing)
     {
-        pixelPrologue << "    gl_FrontFacing = isFrontFace;\n";
+        if (shaderModel <= 3)
+        {
+            pixelPrologue << "    gl_FrontFacing = (vFace * dx_DepthFront.z >= 0.0);\n";
+        }
+        else
+        {
+            pixelPrologue << "    gl_FrontFacing = isFrontFace;\n";
+        }
     }
 
     bool declareSampleID = false;
@@ -984,7 +1070,8 @@ void DynamicHLSL::GenerateShaderLinkHLSL(
 
         if (fragmentShaderD3D->usesFrontFacing)
         {
-            pixelMainParametersStream << ", bool isFrontFace : SV_IsFrontFace";
+            pixelMainParametersStream << (shaderModel >= 4 ? ", bool isFrontFace : SV_IsFrontFace"
+                                                           : ", float vFace : VFACE");
         }
 
         if (declareSampleID)
@@ -1009,21 +1096,25 @@ void DynamicHLSL::GenerateShaderLinkHLSL(
 }
 
 // static
-std::string DynamicHLSL::GenerateGeometryShaderPreamble(const VaryingPacking &varyingPacking,
+std::string DynamicHLSL::GenerateGeometryShaderPreamble(RendererD3D *renderer,
+                                                        const VaryingPacking &varyingPacking,
                                                         const BuiltinVaryingsD3D &builtinsD3D,
                                                         const bool hasMultiviewEnabled,
                                                         const bool selectViewInVS)
 {
+    ASSERT(renderer->getMajorShaderModel() >= 4);
+
     std::ostringstream preambleStream;
 
     const auto &vertexBuiltins = builtinsD3D[gl::ShaderType::Vertex];
 
     preambleStream << "struct GS_INPUT\n";
-    GenerateVaryingLinkHLSL(varyingPacking, vertexBuiltins, FragDepthUsage::Unused, preambleStream);
+    GenerateVaryingLinkHLSL(renderer, varyingPacking, vertexBuiltins, FragDepthUsage::Unused,
+                            builtinsD3D.usesPointSize(), preambleStream);
     preambleStream << "\n"
                       "struct GS_OUTPUT\n";
-    GenerateVaryingLinkHLSL(varyingPacking, builtinsD3D[gl::ShaderType::Geometry],
-                            FragDepthUsage::Unused, preambleStream);
+    GenerateVaryingLinkHLSL(renderer, varyingPacking, builtinsD3D[gl::ShaderType::Geometry],
+                            FragDepthUsage::Unused, builtinsD3D.usesPointSize(), preambleStream);
     preambleStream
         << "\n"
         << "void copyVertex(inout GS_OUTPUT output, GS_INPUT input, GS_INPUT flatinput)\n"
@@ -1085,7 +1176,8 @@ std::string DynamicHLSL::GenerateGeometryShaderPreamble(const VaryingPacking &va
 }
 
 // static
-std::string DynamicHLSL::GenerateGeometryShaderHLSL(const gl::Caps &caps,
+std::string DynamicHLSL::GenerateGeometryShaderHLSL(RendererD3D *renderer,
+                                                    const gl::Caps &caps,
                                                     gl::PrimitiveMode primitiveType,
                                                     const bool useViewScale,
                                                     const bool hasMultiviewEnabled,
@@ -1093,6 +1185,8 @@ std::string DynamicHLSL::GenerateGeometryShaderHLSL(const gl::Caps &caps,
                                                     const bool pointSpriteEmulation,
                                                     const std::string &preambleString)
 {
+    ASSERT(renderer->getMajorShaderModel() >= 4);
+
     std::stringstream shaderStream;
 
     const bool pointSprites = (primitiveType == gl::PrimitiveMode::Points) && pointSpriteEmulation;
@@ -1457,19 +1551,23 @@ BuiltinVaryingsD3D::BuiltinVaryingsD3D(const ProgramD3DMetadata &metadata,
 {
     updateBuiltins(gl::ShaderType::Vertex, metadata, packing);
     updateBuiltins(gl::ShaderType::Fragment, metadata, packing);
-    updateBuiltins(gl::ShaderType::Geometry, metadata, packing);
-    // Some builtins need to be the same in vertex and pixel shaders - input struct needs to be a
-    // prefix of output struct.
-    ASSERT(mBuiltinInfo[gl::ShaderType::Vertex].glPosition.enabled ==
-           mBuiltinInfo[gl::ShaderType::Fragment].glPosition.enabled);
-    ASSERT(mBuiltinInfo[gl::ShaderType::Vertex].glFragCoord.enabled ==
-           mBuiltinInfo[gl::ShaderType::Fragment].glFragCoord.enabled);
-    ASSERT(mBuiltinInfo[gl::ShaderType::Vertex].glPointCoord.enabled ==
-           mBuiltinInfo[gl::ShaderType::Fragment].glPointCoord.enabled);
-    ASSERT(mBuiltinInfo[gl::ShaderType::Vertex].glPointSize.enabled ==
-           mBuiltinInfo[gl::ShaderType::Fragment].glPointSize.enabled);
-    ASSERT(mBuiltinInfo[gl::ShaderType::Vertex].glViewIDOVR.enabled ==
-           mBuiltinInfo[gl::ShaderType::Fragment].glViewIDOVR.enabled);
+    int shaderModel = metadata.getRendererMajorShaderModel();
+    if (shaderModel >= 4)
+    {
+        updateBuiltins(gl::ShaderType::Geometry, metadata, packing);
+    }
+    // In shader model >= 4, some builtins need to be the same in vertex and pixel shaders - input
+    // struct needs to be a prefix of output struct.
+    ASSERT(shaderModel < 4 || mBuiltinInfo[gl::ShaderType::Vertex].glPosition.enabled ==
+                                  mBuiltinInfo[gl::ShaderType::Fragment].glPosition.enabled);
+    ASSERT(shaderModel < 4 || mBuiltinInfo[gl::ShaderType::Vertex].glFragCoord.enabled ==
+                                  mBuiltinInfo[gl::ShaderType::Fragment].glFragCoord.enabled);
+    ASSERT(shaderModel < 4 || mBuiltinInfo[gl::ShaderType::Vertex].glPointCoord.enabled ==
+                                  mBuiltinInfo[gl::ShaderType::Fragment].glPointCoord.enabled);
+    ASSERT(shaderModel < 4 || mBuiltinInfo[gl::ShaderType::Vertex].glPointSize.enabled ==
+                                  mBuiltinInfo[gl::ShaderType::Fragment].glPointSize.enabled);
+    ASSERT(shaderModel < 4 || mBuiltinInfo[gl::ShaderType::Vertex].glViewIDOVR.enabled ==
+                                  mBuiltinInfo[gl::ShaderType::Fragment].glViewIDOVR.enabled);
 }
 
 BuiltinVaryingsD3D::~BuiltinVaryingsD3D() = default;
@@ -1478,17 +1576,29 @@ void BuiltinVaryingsD3D::updateBuiltins(gl::ShaderType shaderType,
                                         const ProgramD3DMetadata &metadata,
                                         const VaryingPacking &packing)
 {
-    const std::string userSemantic = "TEXCOORD";
+    const std::string &userSemantic = GetVaryingSemantic(metadata.getRendererMajorShaderModel(),
+                                                         metadata.usesSystemValuePointSize());
 
-    // Note that when enabling builtins only for specific shader stages, the code needs to ensure
-    // that the input struct of the shader stage is a prefix of the output struct of the previous
-    // stage.
+    // Note that when enabling builtins only for specific shader stages in shader model >= 4, the
+    // code needs to ensure that the input struct of the shader stage is a prefix of the output
+    // struct of the previous stage.
 
     unsigned int reservedSemanticIndex = packing.getMaxSemanticIndex();
 
     BuiltinInfo *builtins = &mBuiltinInfo[shaderType];
 
-    builtins->dxPosition.enableSystem("SV_Position");
+    if (metadata.getRendererMajorShaderModel() >= 4)
+    {
+        builtins->dxPosition.enableSystem("SV_Position");
+    }
+    else if (shaderType == gl::ShaderType::Fragment)
+    {
+        builtins->dxPosition.enableSystem("VPOS");
+    }
+    else
+    {
+        builtins->dxPosition.enableSystem("POSITION");
+    }
 
     if (metadata.usesTransformFeedbackGLPosition())
     {
@@ -1515,8 +1625,16 @@ void BuiltinVaryingsD3D::updateBuiltins(gl::ShaderType shaderType,
     if (shaderType == gl::ShaderType::Vertex ? metadata.addsPointCoordToVertexShader()
                                              : metadata.usesPointCoord())
     {
+        // SM3 reserves the TEXCOORD semantic for point sprite texcoords (gl_PointCoord)
         // In D3D11 we manually compute gl_PointCoord in the GS.
-        builtins->glPointCoord.enable(userSemantic, reservedSemanticIndex++);
+        if (metadata.getRendererMajorShaderModel() >= 4)
+        {
+            builtins->glPointCoord.enable(userSemantic, reservedSemanticIndex++);
+        }
+        else
+        {
+            builtins->glPointCoord.enable("TEXCOORD", 0);
+        }
     }
 
     if (metadata.hasMultiviewEnabled())
@@ -1532,7 +1650,9 @@ void BuiltinVaryingsD3D::updateBuiltins(gl::ShaderType shaderType,
         }
     }
 
-    if (metadata.usesSystemValuePointSize())
+    // Special case: do not include PSIZE semantic in HLSL 3 pixel shaders
+    if (metadata.usesSystemValuePointSize() &&
+        (shaderType != gl::ShaderType::Fragment || metadata.getRendererMajorShaderModel() >= 4))
     {
         builtins->glPointSize.enableSystem("PSIZE");
     }
diff --git a/src/libANGLE/renderer/d3d/DynamicHLSL.h b/src/libANGLE/renderer/d3d/DynamicHLSL.h
index 7cd9f5e62e..29c0fd5b06 100644
--- a/third_party/angle/src/libANGLE/renderer/d3d/DynamicHLSL.h
+++ b/third_party/angle/src/libANGLE/renderer/d3d/DynamicHLSL.h
@@ -114,6 +114,13 @@ struct BuiltinInfo
     BuiltinVarying glLayer;
 };
 
+inline std::string GetVaryingSemantic(int majorShaderModel, bool programUsesPointSize)
+{
+    // SM3 reserves the TEXCOORD semantic for point sprite texcoords (gl_PointCoord)
+    // In D3D11 we manually compute gl_PointCoord in the GS.
+    return ((programUsesPointSize && majorShaderModel < 4) ? "COLOR" : "TEXCOORD");
+}
+
 class BuiltinVaryingsD3D
 {
   public:
@@ -146,6 +153,7 @@ class DynamicHLSL : angle::NonCopyable
         const std::vector<gl::ProgramInput> &shaderAttributes,
         size_t baseUAVRegister);
     static std::string GeneratePixelShaderForOutputSignature(
+        RendererD3D *renderer,
         const std::string &sourceShader,
         const std::vector<PixelShaderOutputVariable> &outputVariables,
         FragDepthUsage fragDepthUsage,
@@ -161,6 +169,7 @@ class DynamicHLSL : angle::NonCopyable
         const gl::ImageUnitTextureTypeMap &image2DBindLayout,
         unsigned int baseUAVRegister);
     static void GenerateShaderLinkHLSL(
+        RendererD3D *renderer,
         const gl::Caps &caps,
         const gl::ShaderMap<gl::SharedCompiledShaderState> &shaderData,
         const gl::ShaderMap<SharedCompiledShaderStateD3D> &shaderDataD3D,
@@ -169,12 +178,14 @@ class DynamicHLSL : angle::NonCopyable
         const BuiltinVaryingsD3D &builtinsD3D,
         gl::ShaderMap<std::string> *shaderHLSL);
 
-    static std::string GenerateGeometryShaderPreamble(const gl::VaryingPacking &varyingPacking,
+    static std::string GenerateGeometryShaderPreamble(RendererD3D *renderer,
+                                                      const gl::VaryingPacking &varyingPacking,
                                                       const BuiltinVaryingsD3D &builtinsD3D,
                                                       const bool hasMultiviewEnabled,
                                                       const bool selectViewInVS);
 
-    static std::string GenerateGeometryShaderHLSL(const gl::Caps &caps,
+    static std::string GenerateGeometryShaderHLSL(RendererD3D *renderer,
+                                                  const gl::Caps &caps,
                                                   gl::PrimitiveMode primitiveType,
                                                   const bool useViewScale,
                                                   const bool hasMultiviewEnabled,
@@ -190,9 +201,11 @@ class DynamicHLSL : angle::NonCopyable
                                         std::vector<PixelShaderOutputVariable> *outPixelShaderKey);
 
   private:
-    static void GenerateVaryingLinkHLSL(const gl::VaryingPacking &varyingPacking,
+    static void GenerateVaryingLinkHLSL(RendererD3D *renderer,
+                                        const gl::VaryingPacking &varyingPacking,
                                         const BuiltinInfo &builtins,
                                         FragDepthUsage fragDepthUsage,
+                                        bool programUsesPointSize,
                                         std::ostringstream &hlslStream);
 
     static void GenerateAttributeConversionHLSL(angle::FormatID vertexFormatID,
diff --git a/src/libANGLE/renderer/d3d/ProgramD3D.cpp b/src/libANGLE/renderer/d3d/ProgramD3D.cpp
index 3673e814e4..a5eb2253c0 100644
--- a/third_party/angle/src/libANGLE/renderer/d3d/ProgramD3D.cpp
+++ b/third_party/angle/src/libANGLE/renderer/d3d/ProgramD3D.cpp
@@ -178,7 +178,8 @@ ProgramD3DMetadata::ProgramD3DMetadata(
     const gl::SharedCompiledShaderState &fragmentShader,
     const gl::ShaderMap<SharedCompiledShaderStateD3D> &attachedShaders,
     int shaderVersion)
-    : mUsesViewScale(renderer->presentPathFastEnabled()),
+    : mRendererMajorShaderModel(renderer->getMajorShaderModel()),
+      mUsesViewScale(renderer->presentPathFastEnabled()),
       mCanSelectViewInVertexShader(renderer->canSelectViewInVertexShader()),
       mFragmentShader(fragmentShader),
       mAttachedShaders(attachedShaders),
@@ -187,6 +188,11 @@ ProgramD3DMetadata::ProgramD3DMetadata(
 
 ProgramD3DMetadata::~ProgramD3DMetadata() = default;
 
+int ProgramD3DMetadata::getRendererMajorShaderModel() const
+{
+    return mRendererMajorShaderModel;
+}
+
 bool ProgramD3DMetadata::usesBroadcast(const gl::Version &clientVersion) const
 {
     const SharedCompiledShaderStateD3D &shader = mAttachedShaders[gl::ShaderType::Fragment];
@@ -226,7 +232,7 @@ bool ProgramD3DMetadata::usesPointSize() const
 
 bool ProgramD3DMetadata::usesInsertedPointCoordValue() const
 {
-    return usesPointCoord();
+    return usesPointCoord() && mRendererMajorShaderModel >= 4;
 }
 
 bool ProgramD3DMetadata::usesViewScale() const
@@ -649,7 +655,7 @@ angle::Result ProgramD3D::linkJobImpl(d3d::Context *context,
                                 mState.getAttachedShader(gl::ShaderType::Vertex)->shaderVersion);
     BuiltinVaryingsD3D builtins(metadata, varyingPacking);
 
-    DynamicHLSL::GenerateShaderLinkHLSL(caps, mState.getAttachedShaders(),
+    DynamicHLSL::GenerateShaderLinkHLSL(mRenderer, caps, mState.getAttachedShaders(),
                                         executableD3D->mAttachedShaders, metadata, varyingPacking,
                                         builtins, &executableD3D->mShaderHLSL);
 
@@ -668,16 +674,19 @@ angle::Result ProgramD3D::linkJobImpl(d3d::Context *context,
     executableD3D->mUsesFlatInterpolation =
         FindFlatInterpolationVarying(mState.getAttachedShaders());
 
-    executableD3D->mGeometryShaderPreamble = DynamicHLSL::GenerateGeometryShaderPreamble(
-        varyingPacking, builtins, executableD3D->mHasMultiviewEnabled,
-        metadata.canSelectViewInVertexShader());
+    if (mRenderer->getMajorShaderModel() >= 4)
+    {
+        executableD3D->mGeometryShaderPreamble = DynamicHLSL::GenerateGeometryShaderPreamble(
+            mRenderer, varyingPacking, builtins, executableD3D->mHasMultiviewEnabled,
+            metadata.canSelectViewInVertexShader());
+    }
 
     executableD3D->initAttribLocationsToD3DSemantic(
         mState.getAttachedShader(gl::ShaderType::Vertex));
 
     executableD3D->defineUniformsAndAssignRegisters(mRenderer, mState.getAttachedShaders());
 
-    executableD3D->gatherTransformFeedbackVaryings(varyingPacking,
+    executableD3D->gatherTransformFeedbackVaryings(mRenderer, varyingPacking,
                                                    mState.getTransformFeedbackVaryingNames(),
                                                    builtins[gl::ShaderType::Vertex]);
 
diff --git a/src/libANGLE/renderer/d3d/ProgramD3D.h b/src/libANGLE/renderer/d3d/ProgramD3D.h
index bdbed75708..48cf677c43 100644
--- a/third_party/angle/src/libANGLE/renderer/d3d/ProgramD3D.h
+++ b/third_party/angle/src/libANGLE/renderer/d3d/ProgramD3D.h
@@ -34,6 +34,7 @@ class ProgramD3DMetadata final : angle::NonCopyable
                        int shaderVersion);
     ~ProgramD3DMetadata();
 
+    int getRendererMajorShaderModel() const;
     bool usesBroadcast(const gl::Version &clientVersion) const;
     bool usesSecondaryColor() const;
     bool usesPointCoord() const;
@@ -57,6 +58,7 @@ class ProgramD3DMetadata final : angle::NonCopyable
     uint8_t getCullDistanceArraySize() const;
 
   private:
+    const int mRendererMajorShaderModel;
     const bool mUsesViewScale;
     const bool mCanSelectViewInVertexShader;
     gl::SharedCompiledShaderState mFragmentShader;
diff --git a/src/libANGLE/renderer/d3d/ProgramExecutableD3D.cpp b/src/libANGLE/renderer/d3d/ProgramExecutableD3D.cpp
index 01894182f9..506eabdb4f 100644
--- a/third_party/angle/src/libANGLE/renderer/d3d/ProgramExecutableD3D.cpp
+++ b/third_party/angle/src/libANGLE/renderer/d3d/ProgramExecutableD3D.cpp
@@ -360,15 +360,11 @@ void D3DVertexExecutable::getSignature(RendererD3D *renderer,
     {
         angle::FormatID vertexFormatID = inputLayout[index];
         if (vertexFormatID == angle::FormatID::NONE)
-        {
             continue;
-        }
 
         VertexConversionType conversionType = renderer->getVertexConversionType(vertexFormatID);
         if ((conversionType & VERTEX_CONVERT_GPU) == 0)
-        {
             continue;
-        }
 
         GLenum componentType   = renderer->getVertexComponentType(vertexFormatID);
         (*signatureOut)[index] = GetAttribType(componentType);
@@ -384,9 +380,7 @@ bool D3DVertexExecutable::matchesSignature(const Signature &signature) const
         auto a = index < signature.size() ? signature[index] : HLSLAttribType::FLOAT;
         auto b = index < mSignature.size() ? mSignature[index] : HLSLAttribType::FLOAT;
         if (a != b)
-        {
             return false;
-        }
     }
 
     return true;
@@ -532,7 +526,7 @@ bool ProgramExecutableD3D::load(const gl::Context *context,
 
     {
         const gl::ShaderType shaderType = gl::ShaderType::Fragment;
-        size_t imageCount               = stream->readInt<size_t>();
+        size_t imageCount = stream->readInt<size_t>();
         for (size_t imageIndex = 0; imageIndex < imageCount; ++imageIndex)
         {
             D3DImage image;
@@ -568,7 +562,7 @@ bool ProgramExecutableD3D::load(const gl::Context *context,
 
     {
         const gl::ShaderType shaderType = gl::ShaderType::Fragment;
-        size_t image2DUniformCount      = stream->readInt<size_t>();
+        size_t image2DUniformCount = stream->readInt<size_t>();
         if (stream->error())
         {
             infoLog << "Invalid program binary.";
@@ -702,7 +696,7 @@ angle::Result ProgramExecutableD3D::loadBinaryShaderExecutables(d3d::Context *co
 
         size_t vertexShaderSize                   = stream->readInt<size_t>();
         const unsigned char *vertexShaderFunction = stream->remainingSpan().data();
-        ShaderExecutableD3D *shaderExecutable     = nullptr;
+        ShaderExecutableD3D *shaderExecutable = nullptr;
 
         ANGLE_TRY(renderer->loadExecutable(contextD3D, vertexShaderFunction, vertexShaderSize,
                                            gl::ShaderType::Vertex, mStreamOutVaryings,
@@ -1107,9 +1101,7 @@ void ProgramExecutableD3D::updateSamplerMapping()
     for (const D3DUniform *d3dUniform : mD3DUniforms)
     {
         if (!d3dUniform->isSampler())
-        {
             continue;
-        }
 
         int count = d3dUniform->getArraySizeProduct();
 
@@ -1167,14 +1159,14 @@ gl::RangeUI ProgramExecutableD3D::getUsedImageRange(gl::ShaderType type, bool re
     return readonly ? mUsedReadonlyImageRange[type] : mUsedImageRange[type];
 }
 
-bool ProgramExecutableD3D::usesPointSpriteEmulation() const
+bool ProgramExecutableD3D::usesPointSpriteEmulation(RendererD3D *renderer) const
 {
-    return mUsesPointSize;
+    return mUsesPointSize && renderer->getMajorShaderModel() >= 4;
 }
 
-bool ProgramExecutableD3D::usesGeometryShaderForPointSpriteEmulation() const
+bool ProgramExecutableD3D::usesGeometryShaderForPointSpriteEmulation(RendererD3D *renderer) const
 {
-    return usesPointSpriteEmulation();
+    return usesPointSpriteEmulation(renderer);
 }
 
 bool ProgramExecutableD3D::usesGeometryShader(RendererD3D *renderer,
@@ -1193,7 +1185,7 @@ bool ProgramExecutableD3D::usesGeometryShader(RendererD3D *renderer,
         }
         return provokingVertex == gl::ProvokingVertexConvention::LastVertexConvention;
     }
-    return usesGeometryShaderForPointSpriteEmulation();
+    return usesGeometryShaderForPointSpriteEmulation(renderer);
 }
 
 angle::Result ProgramExecutableD3D::getVertexExecutableForCachedInputLayout(
@@ -1277,9 +1269,9 @@ angle::Result ProgramExecutableD3D::getGeometryExecutableForPrimitiveType(
     }
 
     std::string geometryHLSL = DynamicHLSL::GenerateGeometryShaderHLSL(
-        caps, geometryShaderType, renderer->presentPathFastEnabled(), mHasMultiviewEnabled,
-        renderer->canSelectViewInVertexShader(), usesGeometryShaderForPointSpriteEmulation(),
-        mGeometryShaderPreamble);
+        renderer, caps, geometryShaderType, renderer->presentPathFastEnabled(),
+        mHasMultiviewEnabled, renderer->canSelectViewInVertexShader(),
+        usesGeometryShaderForPointSpriteEmulation(renderer), mGeometryShaderPreamble);
 
     gl::InfoLog tempInfoLog;
     gl::InfoLog *currentInfoLog = infoLog ? infoLog : &tempInfoLog;
@@ -1321,8 +1313,8 @@ angle::Result ProgramExecutableD3D::getPixelExecutableForCachedOutputLayout(
     }
 
     std::string pixelHLSL = DynamicHLSL::GeneratePixelShaderForOutputSignature(
-        mShaderHLSL[gl::ShaderType::Fragment], mPixelShaderKey, mFragDepthUsage, mUsesSampleMask,
-        mPixelShaderOutputLayoutCache, mPixelShaderKey.size());
+        renderer, mShaderHLSL[gl::ShaderType::Fragment], mPixelShaderKey, mFragDepthUsage,
+        mUsesSampleMask, mPixelShaderOutputLayoutCache, mPixelShaderKey.size());
 
     std::string finalPixelHLSL = DynamicHLSL::GenerateShaderForImage2DBindSignature(
         *this, gl::ShaderType::Fragment, mAttachedShaders[gl::ShaderType::Fragment], pixelHLSL,
@@ -1744,9 +1736,7 @@ void ProgramExecutableD3D::defineUniformsAndAssignRegisters(
     {
         const gl::LinkedUniform &glUniform = mExecutable->getUniforms()[index];
         if (!glUniform.isInDefaultBlock())
-        {
             continue;
-        }
 
         std::string name = mExecutable->getUniformNames()[index];
         if (glUniform.isArray())
@@ -1975,11 +1965,13 @@ void ProgramExecutableD3D::assignImage2DRegisters(gl::ShaderType shaderType,
 }
 
 void ProgramExecutableD3D::gatherTransformFeedbackVaryings(
+    RendererD3D *renderer,
     const gl::VaryingPacking &varyingPacking,
     const std::vector<std::string> &tfVaryingNames,
     const BuiltinInfo &builtins)
 {
-    const std::string varyingSemantic = "TEXCOORD";
+    const std::string &varyingSemantic =
+        GetVaryingSemantic(renderer->getMajorShaderModel(), usesPointSize());
 
     // Gather the linked varyings that are used for transform feedback, they should all exist.
     mStreamOutVaryings.clear();
diff --git a/src/libANGLE/renderer/d3d/ProgramExecutableD3D.h b/src/libANGLE/renderer/d3d/ProgramExecutableD3D.h
index cbbe24332a..ad9cf340fd 100644
--- a/third_party/angle/src/libANGLE/renderer/d3d/ProgramExecutableD3D.h
+++ b/third_party/angle/src/libANGLE/renderer/d3d/ProgramExecutableD3D.h
@@ -331,11 +331,11 @@ class ProgramExecutableD3D : public ProgramExecutableImpl
     gl::RangeUI getUsedImageRange(gl::ShaderType type, bool readonly) const;
 
     bool usesPointSize() const { return mUsesPointSize; }
-    bool usesPointSpriteEmulation() const;
+    bool usesPointSpriteEmulation(RendererD3D *renderer) const;
     bool usesGeometryShader(RendererD3D *renderer,
                             const gl::ProvokingVertexConvention provokingVertex,
                             gl::PrimitiveMode drawMode) const;
-    bool usesGeometryShaderForPointSpriteEmulation() const;
+    bool usesGeometryShaderForPointSpriteEmulation(RendererD3D *renderer) const;
 
     angle::Result getVertexExecutableForCachedInputLayout(d3d::Context *context,
                                                           RendererD3D *renderer,
@@ -430,6 +430,7 @@ class ProgramExecutableD3D : public ProgramExecutableImpl
                              gl::RangeUI *outUsedRange);
 
     void gatherTransformFeedbackVaryings(
+        RendererD3D *renderer,
         const gl::VaryingPacking &varyings,
         const std::vector<std::string> &transformFeedbackVaryingNames,
         const BuiltinInfo &builtins);
diff --git a/src/libANGLE/renderer/d3d/d3d11/StateManager11.cpp b/src/libANGLE/renderer/d3d/d3d11/StateManager11.cpp
index d81a5c7882..fc97c62cbb 100644
--- a/third_party/angle/src/libANGLE/renderer/d3d/d3d11/StateManager11.cpp
+++ b/third_party/angle/src/libANGLE/renderer/d3d/d3d11/StateManager11.cpp
@@ -162,6 +162,11 @@ void UpdateUniformBuffer(ID3D11DeviceContext *deviceContext,
                                      0);
 }
 
+size_t GetReservedBufferCount(bool usesPointSpriteEmulation)
+{
+    return usesPointSpriteEmulation ? 1 : 0;
+}
+
 bool CullsEverything(const gl::State &glState)
 {
     return (glState.getRasterizerState().cullFace &&
@@ -356,6 +361,13 @@ void ShaderConstants11::onViewportChange(const gl::Rectangle &glViewport,
     mPixel.viewCoords[2] = glViewport.x + (glViewport.width * 0.5f);
     mPixel.viewCoords[3] = glViewport.y + (glViewport.height * 0.5f);
 
+    // Instanced pointsprite emulation requires ViewCoords to be defined in the
+    // the vertex shader.
+    mVertex.viewCoords[0] = mPixel.viewCoords[0];
+    mVertex.viewCoords[1] = mPixel.viewCoords[1];
+    mVertex.viewCoords[2] = mPixel.viewCoords[2];
+    mVertex.viewCoords[3] = mPixel.viewCoords[3];
+
     const float zNear = dxViewport.MinDepth;
     const float zFar  = dxViewport.MaxDepth;
 
@@ -1212,7 +1224,9 @@ void StateManager11::syncViewport(const gl::Context *context)
     {
         // When present path fast is active and we're rendering to framebuffer 0, we must invert
         // the viewport in Y-axis.
-        // NOTE: We delay the inversion until right before the call to RSSetViewports.
+        // NOTE: We delay the inversion until right before the call to RSSetViewports, and leave
+        // dxViewportTopLeftY unchanged. This allows us to calculate viewAdjust below using the
+        // unaltered dxViewportTopLeftY value.
         dxViewport.TopLeftY = static_cast<float>(mCurPresentPathFastColorBufferHeight -
                                                  dxViewportTopLeftY - dxViewportHeight);
     }
@@ -1779,9 +1793,7 @@ angle::Result StateManager11::syncCurrentValueAttribs(
     for (auto attribIndex : dirtyActiveAttribs)
     {
         if (vertexAttributes[attribIndex].enabled)
-        {
             continue;
-        }
 
         const auto *attrib                      = &vertexAttributes[attribIndex];
         const auto &currentValue                = currentValues[attribIndex];
@@ -2813,6 +2825,36 @@ bool StateManager11::syncIndexBuffer(ID3D11Buffer *buffer,
     return false;
 }
 
+// Vertex buffer is invalidated outside this function.
+angle::Result StateManager11::updateVertexOffsetsForPointSpritesEmulation(
+    const gl::Context *context,
+    GLint startVertex,
+    GLsizei emulatedInstanceId)
+{
+    size_t reservedBuffers = GetReservedBufferCount(true);
+    for (size_t attribIndex = 0; attribIndex < mCurrentAttributes.size(); ++attribIndex)
+    {
+        const auto &attrib = *mCurrentAttributes[attribIndex];
+        size_t bufferIndex = reservedBuffers + attribIndex;
+
+        if (attrib.divisor > 0)
+        {
+            unsigned int offset = 0;
+            ANGLE_TRY(attrib.computeOffset(context, startVertex, &offset));
+            offset += (attrib.stride * (emulatedInstanceId / attrib.divisor));
+            if (offset != mCurrentVertexOffsets[bufferIndex])
+            {
+                invalidateInputLayout();
+                mDirtyVertexBufferRange.extend(static_cast<unsigned int>(bufferIndex));
+                mCurrentVertexOffsets[bufferIndex] = offset;
+            }
+        }
+    }
+
+    applyVertexBufferChanges();
+    return angle::Result::Continue;
+}
+
 angle::Result StateManager11::generateSwizzle(const gl::Context *context, gl::Texture *texture)
 {
     if (!texture)
diff --git a/src/libANGLE/renderer/d3d/d3d11/StateManager11.h b/src/libANGLE/renderer/d3d/d3d11/StateManager11.h
index 619e718467..5d0f3d015c 100644
--- a/third_party/angle/src/libANGLE/renderer/d3d/d3d11/StateManager11.h
+++ b/third_party/angle/src/libANGLE/renderer/d3d/d3d11/StateManager11.h
@@ -64,6 +64,8 @@ class ShaderConstants11 : angle::NonCopyable
     {
         Vertex()
             : depthRange{.0f},
+              viewAdjust{.0f},
+              viewCoords{.0f},
               viewScale{.0f},
               clipControlOrigin{-1.0f},
               clipControlZeroToOne{.0f},
@@ -73,6 +75,8 @@ class ShaderConstants11 : angle::NonCopyable
         {}
 
         float depthRange[4];
+        float viewAdjust[4];
+        float viewCoords[4];
         float viewScale[2];
 
         // EXT_clip_control
@@ -170,7 +174,7 @@ class StateManager11 final : angle::NonCopyable
     // The Context is allowed to be nullptr for these methods, when called in EGL init code.
     void invalidateRenderTarget();
 
-    // Invalidates all current vertex buffer bindings.
+    // Called by instanced point sprite emulation.
     void invalidateVertexBuffer();
 
     // Called by Framebuffer11::syncState for the default sized viewport.
@@ -247,6 +251,10 @@ class StateManager11 final : angle::NonCopyable
 
     void setIndexBuffer(ID3D11Buffer *buffer, DXGI_FORMAT indexFormat, unsigned int offset);
 
+    angle::Result updateVertexOffsetsForPointSpritesEmulation(const gl::Context *context,
+                                                              GLint startVertex,
+                                                              GLsizei emulatedInstanceId);
+
     // Only used in testing.
     InputLayoutCache *getInputLayoutCache() { return &mInputLayoutCache; }
 
